lib/galaxy/webapps/galaxy/controllers/workflow.py

```diff
@@ -432,15 +428,29 @@ def save_workflow_as(
                     workflow_data,
                     workflow_update_options,
                 )
+            except exceptions.DuplicatedIdentifierException as e:
+                return self.message_exception(trans, e.err_msg, False)
             except MissingToolsException as e:
-                return dict(
+                missing_tools_return = dict(
                     name=e.workflow.name,
                     message=(
                         "This workflow includes missing or invalid tools. "
                         "It cannot be saved until the following steps are removed or the missing tools are enabled."
                     ),
                     errors=e.errors,
                 )
+            else:
+                missing_tools_return = None
+
+            # Persist
+            session = trans.sa_session
+            session.add(stored_workflow)
+            with transaction(session):
+                session.commit()
+
+            if missing_tools_return:
+                return missing_tools_return
```

Comment:
Why did you make that change, it doesn't look related or correct.

Reply:
I would prefer that we focus on the bug before refactoring things. I think that's not a clearcut improvement.

Reply:
But in order to fix this bug, we need to catch this exception

Reply:
No, you need to use the API. This used to work fine on 23.1

Reply:
Ok, I'll look at what we did in 23.1, and try to use the api instead of the controller here.

Reply:
What you want to do is replace `save_workflow_as` with a PUT to `/api/workflows/{id}` or a POST to `/api/workflows` (depending on whether the workflow has been saved before).

Reply:
Done, used `POST api/workflows` instead for both creating and saving-as workflows. It works and the errors are correctly handled at the client as well.